src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala

```diff
@@ -43,4 +51,49 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
       WorkspaceAccessLevels.Write -> wGroup,
       WorkspaceAccessLevels.Read -> rGroup)
   }
+
+  //turns a PipeFunction into one that takes a LoopBundle
+  implicit def pipeToLoopBundle[A,B](f: PipeFunction[A,B]) = new PipeFunction[LoopPipe.LoopBundle[A], B] {
+    override def compute(bundle: LoopPipe.LoopBundle[A]) : B = f.compute(bundle.getObject)
+  }
+
+  private def invert[A](f: PipeFunction[A, java.lang.Boolean]) = new PipeFunction[A, java.lang.Boolean] {
+    override def compute(v: A) = !f.compute(v)
+  }
+
+  private def isConnectedToTargetWorkspace(wsId:String) = new PipeFunction[Vertex, java.lang.Boolean] {
+    override def compute(v: Vertex) = v.getVertices(Direction.IN).iterator.exists({ case ws =>
+      isVertexOfClass(VertexSchema.Workspace).compute(ws) &&
+        hasPropertyValue("workspaceId", wsId).compute(ws) })
+  }
+
+  override def getMaximumAccessLevel(userSubjectId: String, workspaceId: String, txn: RawlsTransaction): WorkspaceAccessLevel = {
+    txn withGraph { db =>
+
+      //Returns a list of paths starting with a group the user belongs in and ending with the workspace's access map.
+      // user -> group -> ...some other groups... -> group -> workspace access map
+      //Multiple paths implies that the user belongs to multiple groups associated with the workspace.
+      val accessMapPaths =
+        userPipeline(db, userSubjectId).as("foo").in()
+          .loop(
+            "foo", //start point of loop
```

Comment:
The user isn't always the "current" node in the loop condition, but I'll make it a different word regardless. Maybe bar.